client/transaction-pool/graph/src/watcher.rs

```diff
@@ -85,26 +85,41 @@ impl<H: Clone, H2: Clone> Sender<H, H2> {
 	/// Some state change (perhaps another extrinsic was included) rendered this extrinsic invalid.
 	pub fn usurped(&mut self, hash: H) {
 		self.send(TransactionStatus::Usurped(hash));
-		self.finalized = true;
+		self.is_finalized = true;
 	}
 
 	/// Extrinsic has been included in block with given hash.
-	pub fn in_block(&mut self, hash: H2) {
+	pub fn in_block(&mut self, hash: BH) {
 		self.send(TransactionStatus::InBlock(hash));
-		self.finalized = true;
+	}
+
+	/// Extrinsic has been finalized by a finality gadget.
+	pub fn finalized(&mut self, hash: &BH) {
+		self.send(TransactionStatus::Finalized(hash.clone()));
+		self.is_finalized = true;
+	}
+
+	/// The block this extrinsic was included in has been retracted
+	pub fn finality_timeout(&mut self, hash: &BH) {
+		self.send(TransactionStatus::FinalityTimeout(hash.clone()));
+	}
+
+	/// The block this extrinsic was included in has been retracted
+	pub fn retracted(&mut self, hash: &BH) {
```

Comment:
```suggestion
	pub fn retracted(&mut self, hash: BH) {
```
and here. It's better to make the caller aware that you need an owned hash, than to pretend you are fine with a reference and then clone internally.